### DISCUSSION TITLE
Proper guide for OAuth for React Native?

### DISCUSSION THREAD

**Discussion by shaun49995** [pocketbase/pocketbase#5229] [Q&A]
We struggle to get Google Auth working in our react native expo app. We've implemented quite a few solutions found in some of the GitHub threads, and got it to work on iOS but not Android. ios also does not auto-close after auth and is very buggy. Is there a complete guide or code that we can use?

Help would be appreciated.

**Comment by ganigeorgiev** [comment]
There is no official step-by-step guide that will cover everything and to be honest I don't think there is need for it.

If you want a minimal code sample you can check https://snack.expo.dev/_y-uEipvIMWI68vl2gKhF (it is for an older Expo version, but everything else should be the same).


By not elaborating and just saying _"very buggy"_ I don't know how to help.

Is there an error thrown when attempting the authentication process (you could wrap your auth call in `try/catch` and `console.log(err.response, err.originalError)`)? 

How do you even open the url? What SSE polyfill do you use? Could you provide a **minimal** [snack.expo.dev example](https://snack.expo.dev/) with your failing code?

Since you are having troubles with Android, please read https://github.com/pocketbase/pocketbase/discussions/2484#discussioncomment-6118501. The issue on Android is most likely caused by the Flipper debug plugin and I don't think there is anything we can do from the PocketBase JS SDK side to fix it (_even removing the dependency on the native JS `EventSource` is not a solution at the moment because the last time I tested streamed `fetch` responses didn't work on ReactNative without polyfills_).

**Comment by PalomboDev** [comment]
After a lot of trial/error and reading through the GitHub, I figured out a solution that met my needs. This solution achieves an in-app browser to do the sign-in flow with oAuth App. Once you have signed in successfully with the oAuth App, the in-app browser automatically closes. There is also support for persistent PocketBase data by creating an abstraction off of BaseAuthStore and using @react-native-async-storage/async-storage.

The only provider the below code supports is GitHub (but adding more should be as easy as adding them to the `providers` type). This is the URL I set for "Authorization callback URL":
https://[INSERT POCKETBASE URL]/api/oauth2-redirect

**Expo Version:** ^51.0.24

**PocketBase Code**
https://gist.github.com/PalomboDev/e2d49a9ce993a1193fd77d682959d5f3

_layout.tsx
```
export default function RootLayout() {
 ...

  return (
    <Stack>
      <Stack.Screen
        name="(tabs)"
      />
      <Stack.Screen
        name="(auth)"
      />
    </Stack>
  );
}


```

(tabs)/_layout.tsx
```
export default function TabLayout() {
  const { isAuthenticated } = useAuth();

  if (!isAuthenticated) {
    return <Redirect href="/(auth)/signin" />;
  }

  return (...);
}
```

(auth)/_layout.tsx
```
export default function TabLayout() {
  const { isAuthenticated } = useAuth();

  if (isAuthenticated) {
    return <Redirect href="/(tabs)/" />;
  }

  return (...);
}
```

**Reply by ganigeorgiev** [reply]
_Side-note: I don't entirely understand the above code but note that the JS SDK comes with an AsyncAuthStore out of the box - https://github.com/pocketbase/js-sdk?tab=readme-ov-file#asyncauthstore._

**Reply by PalomboDev** [reply]
I guess all of the GitHub discussions I was reading were outdated. Definitely use the out-the-box solution instead for storage. I've updated the above gist to use the built in AsyncAuthStore.
